bindings/sns/0.2.0/channel.json

```diff
@@ -0,0 +1,200 @@
+  "type": "object",
+  "additionalProperties": false,
+  "patternProperties": {
+    "^x-[\\w\\d\\.\\x2d_]+$": {
+      "$ref": "http://asyncapi.com/definitions/3.0.0/specificationExtension.json"
+    }
+  },
+  "properties": {
+    "name": {
+      "type": "string",
+      "description": "The name of the topic. Can be different from the channel name to allow flexibility around AWS resource naming limitations."
+    },
+    "ordering": {
+      "$ref": "http://asyncapi.com/bindings/sns/0.2.0/channel.json#/definitions/ordering"
+    },
+    "policy": {
+      "$ref": "http://asyncapi.com/bindings/sns/0.2.0/channel.json#/definitions/policy"
+    },
+    "tags": {
+      "type": "object",
+      "description": "Key-value pairs that represent AWS tags on the topic."
+    },
+    "bindingVersion": {
+      "type": "string",
+      "description": "The version of this binding.",
+      "default": "latest"
+    }
+  },
+  "required": [
+    "name"
+  ],
+  "definitions": {
+    "ordering": {
+      "type": "object",
+      "description": "By default, we assume an unordered SNS topic. This field allows configuration of a FIFO SNS Topic.",
+      "patternProperties": {
+        "^x-[\\w\\d\\.\\x2d_]+$": {
+          "$ref": "http://asyncapi.com/definitions/3.0.0/specificationExtension.json"
+        }
+      },
+      "properties": {
+        "type": {
+          "type": "string",
+          "description": "Defines the type of SNS Topic.",
+          "enum": [
+            "standard",
+            "FIFO"
+          ]
+        },
+        "contentBasedDeduplication": {
+          "type": "boolean",
+          "description": "True to turn on de-duplication of messages for a channel."
+        }
+      },
+      "required": [
+        "type"
+      ]
+    },
+    "policy": {
+      "type": "object",
+      "description": "The security policy for the SNS Topic.",
+      "patternProperties": {
+        "^x-[\\w\\d\\.\\x2d_]+$": {
+          "$ref": "http://asyncapi.com/definitions/3.0.0/specificationExtension.json"
+        }
+      },
+      "properties": {
+        "statements": {
+          "type": "array",
+          "description": "An array of statement objects, each of which controls a permission for this topic",
+          "items": {
+            "$ref": "http://asyncapi.com/bindings/sns/0.2.0/channel.json#/definitions/statement"
+          }
+        }
+      },
+      "required": [
+        "statements"
+      ]
+    },
+    "statement": {
+      "type": "object",
+      "patternProperties": {
+        "^x-[\\w\\d\\.\\x2d_]+$": {
+          "$ref": "http://asyncapi.com/definitions/3.0.0/specificationExtension.json"
+        }
+      },
+      "properties": {
+        "effect": {
+          "type": "string",
+          "enum": [
+            "Allow",
+            "Deny"
+          ]
+        },
+        "principal": {
+          "description": "The AWS account(s) or resource ARN(s) that this statement applies to.",
+          "oneOf": [
```

Comment:
Why not just a string array?

Reply:
It can't just be a string array as it needs to be able to take a property and value in some cases https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_principal.html

Reply:
Would it be too much to ask for the objects being defined as in the AWS docs? As OneOf. I think i only see 3 different types. (from quickly glancing)

Reply:
This is what I've attempted to do here, according to the AWS docs the principal property can be:
- a single string value of `"*"`
- a key/value pair of `AWS/Service: {accountId}/{service}`
- a property with an array of string values: `AWS: [123, 456]`

Reply:
from the Docs the actual objects look a bit different.
Youve described them as "any named of pattern property, which is either a string or an array of string.
I would much rather that it is described as its documented by AWS (to leave out any problem with interpretation).
Something like
```
"oneOf": [
  {
    "type": "object",
    "properties": {
      "AWS": {
        "type": "array",
        "items": {
          "type": "string"
        }
      }
    },
    "required": ["AWS"],
    "additionalProperties": false
  },
  {
    "type": "object",
    "properties": {
      "AWS": {
        "type": "string"
      }
    },
    "required": ["AWS"],
    "additionalProperties": false
  },
  {
    "type": "object",
    "properties": {
      "Federated": {
        "type": "string"
      }
    },
    "required": ["Federated"],
    "additionalProperties": false
  },
  {
    "type": "object",
    "properties": {
      "Service": {
        "type": "array",
        "items": {
          "type": "string"
        }
      }
    },
    "required": ["Service"],
    "additionalProperties": false
  },
  {
    "type": "object",
    "properties": {
      "Service": {
        "type": "string"
      }
    },
    "required": ["Service"],
    "additionalProperties": false
  }
]
}
```

Reply:
Yeah that looks good, I'm not sure the `federated` option is needed though? I think that falls under `AWS`? We have a slight issue in that the docs allow just `*` as a value without a property name https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_principal.html#principal-anonymous. It is equivalent to `AWS: *`, so we can enforce that is used instead for the sake of the structure? However, it is technically an option

Reply:
@VisualBean is that a thumbs up to enforcing a property name is always provided, even for the `"*"` case?

Reply:
I think for json and yaml alike `*` is a string either way, so it fits under the 
```
"AWS": {
        "type": "string"
}
```

Reply:
Sorry I mean, to address the fact that AWS allows both:

`Principal: "*"`

and

`Principal: { "AWS" : "*" }`

So I was asking in order to have a consistent structure do we want to enforce a property name (AWS in this case), so we wouldn't accept option one of the above. Does that make sense?

Reply:
I think we need to satisfy the `"*"` option. So going forward we could have:

`"oneOf": [
            {
              "type": "string"
            },
            { 
               "type": "object", 
               "properties": { 
                     "AWS": { 
                          "oneOf": [
                              {
                                  "type": "string"
                              },
                              {
                                   "type": "array",
                                   "items": {
                                        "type": "string"
                                    }
                               }
                             ]
                      } 
                 }, 
                 "required": ["AWS"], 
                 "additionalProperties": false 
            },
            { 
               "type": "object", 
               "properties": { 
                     "Service": { 
                          "oneOf": [
                              {
                                  "type": "string"
                              },
                              {
                                   "type": "array",
                                   "items": {
                                        "type": "string"
                                    }
                               }
                             ]
                      } 
                 }, 
                 "required": ["Service"], 
                 "additionalProperties": false 
            }
          ]`